cmd/speedtest/speedtest.go

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Program Speedtest provides the speedtest command. The reason to keep it seperate from
+// the normal tailscale cli is because it is not yet ready to go in the tailscale binary.
+package main
+
+// Example usage for client command: go run cmd/speedtest/speedtest.go client --d --host 127.0.0.1 --port 8080 --inc 1 --seconds 5
+// This will connect to the server (if one is running) on 127.0.0.1:8080 and start a 5 second speedtest that gathers data in 1 second
+// increments.
+// Example usage for server command: go run cmd/speedtest/speedtest.go server --max-conns 1 --port 8080 --localhost
+// This will start a speedtest server on 127.0.0.1:8080 and allow only 1 concurrent connection at a time.
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"tailscale.com/client/tailscale"
+	"tailscale.com/net/speedtest"
+
+	"github.com/peterbourgon/ff/v2/ffcli"
+)
+
+// Runs the speedtest command as a commandline program
+func main() {
+	args := os.Args[1:]
+	if err := speedtestCmd.Parse(args); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	err := speedtestCmd.Run(context.Background())
+	if err != nil {
+		if err != flag.ErrHelp {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+	}
+}
+
+// Speedtest command that contains the server and client sub commands.
+var speedtestCmd = &ffcli.Command{
+	Name:       "speedtest",
+	ShortUsage: "speedtest {server|client} ...",
+	ShortHelp:  "Run a speed test",
+	Subcommands: []*ffcli.Command{
+		speedtestServerCmd,
+		speedtestClientCmd,
+	},
+	Exec: func(context.Context, []string) error {
+		return errors.New("subcommand required; run 'tailscale speedtest -h' for details")
+	},
+}
+
+// speedtestServerCmd takes necessary info like the port to
+// listen on and then passes them to the StartServer function in the speedtest package.
+// if the localhost flag is given, the server will use 127.0.0.1, otherwise the server will
+// use the tailscale ip address
+var speedtestServerCmd = &ffcli.Command{
+	Name:       "server",
+	ShortUsage: "speedtest server --host <host> --port <port> [--max-conns <max connections>]",
+	ShortHelp:  "Start a speed test server",
+	Exec:       runServer,
+	FlagSet: (func() *flag.FlagSet {
+		fs := flag.NewFlagSet("server", flag.ExitOnError)
+		fs.IntVar(&serverArgs.port, "port", 0, "port to listen on")
+		fs.BoolVar(&serverArgs.localhost, "localhost", false, "use localhost or tailscale ip")
```

Comment:
I imagine we'll want the ability to run the test over Ethernet or Wi-Fi as well to compare throughput on the underlay network versus throughput in the overlay Tailnet.

Having the server listen on ":port", not specifying a specific IP address, would be fine. There wouldn't be a CLI argument then.